src/esw/cameras.py

```diff
@@ -0,0 +1,351 @@
+#!/usr/bin/env python3
+
+from mrover.msg import CameraCmd
+from typing import List, Dict, Tuple
+
+import threading
+
+import rospy
+from mrover.srv import (
+    ChangeCameras,
+    ChangeCamerasRequest,
+    ChangeCamerasResponse,
+)
+
+import jetson.utils
+
+
+class VideoDevices:
+
+    device: int
+    resolution: List[str]
+    video_source: jetson.utils.videoSource
+    # Screens video source is displayed on
+    output_by_endpoint: Dict[str, jetson.utils.videoOutput]
+
+    def __init__(self, device: int):
+        self.resolution = []
+        self.device = device
+        self.video_source = None
+        self.output_by_endpoint = {}
+
+    def remove_endpoint(self, endpoint: str) -> None:
+        """
+        Removes screen endpoint from video source list.
+        If endpoint dictionary is empty, it also deletes video source.
+        :param endpoint: the endpoint (e.g. 10.0.0.7:5000)
+        """
+        assert endpoint in self.output_by_endpoint.keys()
+        del self.output_by_endpoint[endpoint]
+        if len(self.output_by_endpoint) == 0:
+            self.video_source = None
+
+    def create_stream(self, endpoint: str, args: List[str]) -> None:
+        """
+        Adds endpoint to video source list and creates video_source if not existing.
+        :param endpoint: the endpoint (e.g. 10.0.0.7:5000)
+        :param args: a list of strings that are the arguments
+        """
+
+        assert not (endpoint in self.output_by_endpoint.keys())
+
+        if self.is_streaming():
+            # It exists and another stream is using it
+            # TODO - make sure that this check does what we want to do (self.resolution != args)
+            if self.resolution != args:
+                # If different args, just recreate video source and every output
+                try:
+                    # TODO - Make sure that the close does not break anything
+                    self.video_source.Close()
+                    self.video_source = jetson.utils.videoSource(f"/dev/video{self.device}", argv=args)
+                    self.video_source.Open()
+                    for other_endpoint in self.output_by_endpoint.values():
```

Comment:
The way you iterate over output_by_endpoint here doesn't make sense. It seems like you mean to use keys() instead of values() but why not just use items() and get both the key and value at one

Reply:
It was supposed to be key by accident. it was not rigorously tested when changing resolutions so this line of code was never ran.